### PR TITLE
HIVE-28510: Iceberg: FanoutPositionOnlyDeleteWriter support

### PR DESCRIPTION
**Pull request by deniskuzZ**
### What changes were proposed in this pull request?

Adds writer for unordered position deletes. A position delete writer capable of writing to multiple specs and partitions if the incoming stream of deletes is not ordered

### Why are the changes needed?

This writer should allow us to avoid a local sort for delete operations. 

### Does this PR introduce _any_ user-facing change?

No

### Is the change a dependency upgrade?

No

### How was this patch tested?

qtests

### CODE REVIEW

File: iceberg/checkstyle/checkstyle.xml [hunk 1]
```diff
@@ -119,6 +119,7 @@
                 java.util.Collections.*,
                 java.util.stream.Collectors.*,
                 org.apache.commons.lang3.Validate.*,
+                org.apache.hadoop.hive.ql.metadata.VirtualColumn.*,
```
Comment (SourabhBadhya) [inline iceberg/checkstyle/checkstyle.xml:122]: Why this change?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java [hunk 2]
```diff
@@ -350,6 +351,10 @@ public static boolean isCopyOnWriteMode(Context.Operation operation, BinaryOpera
     return RowLevelOperationMode.COPY_ON_WRITE.modeName().equalsIgnoreCase(mode);
   }
 
+  public static boolean isFanoutEnabled(Map<String, String> props) {
+    return PropertyUtil.propertyAsBoolean(props, "write.fanout.enabled", true);
```
Comment (SourabhBadhya) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java:355]: Is `write.fanout.enabled` a new property? Maybe consider creating a String constant.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/writer/HiveIcebergWriterBase.java [hunk 3]
```diff
@@ -77,4 +86,39 @@ protected PartitionKey partition(Record row, int specId) {
...
+
+  // use a fanout writer only if enabled and the input is unordered and the table is partitioned
+  static PartitioningWriter<Record, DataWriteResult> newDataWriter(
+      Table table, HiveFileWriterFactory writers, OutputFileFactory files, Context context) {
+
+    FileIO io = table.io();
+    boolean useFanoutWriter = context.useFanoutWriter();
+    long targetFileSize = context.targetDataFileSize();
+
+    if (table.spec().isPartitioned() && useFanoutWriter) {
```
Comment (SourabhBadhya) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/writer/HiveIcebergWriterBase.java:98]: Fanout writers should be used only when its partitioned? Is this Iceberg / Spark induced behaviour?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/writer/WriterBuilder.java [hunk 4]
```diff
@@ -78,75 +75,58 @@ public WriterBuilder queryId(String newQueryId) {
...
+        .operationId(operationId)
         .build();
 
-    OutputFileFactory deleteOutputFileFactory = OutputFileFactory.builderFor(table, partitionId, taskId)
-        .format(deleteFileFormat)
-        .operationId("delete-" + operationId)
+    OutputFileFactory deleteFileFactory = OutputFileFactory.builderFor(table, partitionId, taskId)
+        .format(context.deleteFileFormat())
+        .operationId(operationId)
+        .suffix("deletes")
```
Comment (SourabhBadhya) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/writer/WriterBuilder.java:109]: Whats this suffix? Is it used in the file name?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputFormat.java [hunk 5]
```diff
@@ -41,10 +41,8 @@
 import org.apache.iceberg.mr.mapred.Container;
 import org.apache.parquet.hadoop.ParquetOutputFormat;
 
-public class HiveIcebergOutputFormat<T> implements OutputFormat<NullWritable, Container<Record>>,
+public class HiveIcebergOutputFormat implements OutputFormat<NullWritable, Container<Record>>,
     HiveOutputFormat<NullWritable, Container<Record>> {
-  private static final String DELETE_FILE_THREAD_POOL_SIZE = "iceberg.delete.file.thread.pool.size";
-  private static final int DELETE_FILE_THREAD_POOL_SIZE_DEFAULT = 10;
```
Comment (SourabhBadhya) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputFormat.java:47]: Are we supporting this config, because removal of these configs might result in backward incompatibility?

File: iceberg/checkstyle/checkstyle.xml (hunk 1 above)
Comment (deniskuzZ) [reply iceberg/checkstyle/checkstyle.xml:122]: to allow static imports

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/writer/HiveIcebergWriterBase.java (hunk 3 above)
Comment (deniskuzZ) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/writer/HiveIcebergWriterBase.java:98]: yes, see SparkPositionDeltaWrite

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/writer/WriterBuilder.java (hunk 4 above)
Comment (deniskuzZ) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/writer/WriterBuilder.java:109]: yes, suffix+operationId

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputFormat.java (hunk 5 above)
Comment (deniskuzZ) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputFormat.java:47]: they are not used anywhere

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java (hunk 2 above)
Comment (deniskuzZ) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/IcebergTableUtil.java]: moved to InputFormatConfig.WRITE_FANOUT_ENABLED

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java [hunk 6]
```diff
@@ -307,27 +307,24 @@ public void commitJobs(List<JobContext> originalContextList, Operation operation
...
 
-  private List<OutputTable> collectOutputs(List<JobContext> jobContextList) {
-    return jobContextList.stream()
-      .flatMap(jobContext -> HiveIcebergStorageHandler.outputTables(jobContext.getJobConf()).stream()
-        .map(output -> new OutputTable(
-          HiveIcebergStorageHandler.catalogName(jobContext.getJobConf(), output),
-          output,
-          SessionStateUtil.getResource(jobContext.getJobConf(), output)
+  private Multimap<OutputTable, JobContext> collectOutputs(List<JobContext> jobContextList) {
+    Multimap<OutputTable, JobContext> outputs =
```
Comment (SourabhBadhya) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java:311]: Why MultiMap? Till now we only considered one JobContext, will this generate multiple JobContexts per OutputTable.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java [hunk 7]
```diff
@@ -346,43 +343,42 @@ public void abortJobs(List<JobContext> originalContextList) throws IOException {
...
-            FilesForCommit results = collectResults(numTasks, fileExecutor, table.location(), jobContext,
-                table.io(), false);
-            // Check if we have files already written and remove data and delta files if there are any
-            Tasks.foreach(results.allFiles())
-                .retry(3)
-                .suppressFailureWhenFinished()
-                .executeWith(fileExecutor)
-                .onFailure((file, exc) -> LOG.warn("Failed to remove data file {} on abort job", file.path(), exc))
-                .run(file -> table.io().deleteFile(file.path().toString()));
+            for (JobContext jobContext : outputs.get(output)) {
```
Comment (SourabhBadhya) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java:364]: Before it was only one JobContext. But now its multiple jobContexts, why?

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java (hunk 6 above)
Comment (deniskuzZ) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java:311]: We have multiple JobContext per OutputTable, nothing changed here, I just used a more convenient collection. Update and Merge operations are represented as multi-insert creating multiple contexts

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java (hunk 7 above)
Comment (deniskuzZ) [reply iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputCommitter.java:364]: it was always like this

File: iceberg/iceberg-handler/src/test/results/positive/merge_iceberg_orc.q.out [hunk 8]
```diff
@@ -68,10 +68,8 @@ STAGE PLANS:
     Tez
 #### A masked pattern was here ####
       Edges:
-        Reducer 2 <- Map 1 (SIMPLE_EDGE), Map 6 (SIMPLE_EDGE)
+        Reducer 2 <- Map 1 (SIMPLE_EDGE), Map 4 (SIMPLE_EDGE)
         Reducer 3 <- Reducer 2 (SIMPLE_EDGE)
-        Reducer 4 <- Reducer 2 (SIMPLE_EDGE)
-        Reducer 5 <- Reducer 2 (SIMPLE_EDGE)
```
Comment (kasakrisz) [inline iceberg/iceberg-handler/src/test/results/positive/merge_iceberg_orc.q.out:74]: Similar to the case when a truncate table is transformed into a delete, no sort reducers are required. I think this change make sense.

**Review by kasakrisz — COMMENTED** [review]
Most plan changes are expected because we do not sort using reducers when writing the delete delta files. IIUC from now it is `FanoutPositionOnlyDeleteWriter`'s responsibility to ensure the correct record order.

File: iceberg/iceberg-handler/src/test/results/positive/merge_iceberg_orc.q.out [hunk 9]
```diff
@@ -186,67 +188,37 @@ STAGE PLANS:
...
                         minReductionHashAggr: 0.4
                         mode: hash
-                        outputColumnNames: _col0, _col1, _col2, _col3, _col4, _col5
-                        Statistics: Num rows: 4 Data size: 1196 Basic stats: COMPLETE Column stats: COMPLETE
+                        outputColumnNames: _col0, _col1, _col2, _col3, _col4
+                        Statistics: Num rows: 4 Data size: 644 Basic stats: COMPLETE Column stats: COMPLETE
                         Reduce Output Operator
-                          key expressions: _col0 (type: int), _col1 (type: bigint), _col2 (type: string), _col3 (type: bigint), _col4 (type: string)
-                          null sort order: zzzzz
-                          sort order: +++++
```
[the remark this answers is not in the document]
Comment (deniskuzZ) [reply iceberg/iceberg-handler/src/test/results/positive/merge_iceberg_orc.q.out:202]: yes

File: ql/src/java/org/apache/hadoop/hive/ql/metadata/HiveStorageHandler.java [hunk 10]
```diff
@@ -498,6 +493,11 @@ default FieldSchema getRowId() {
    * @return the list of columns that should be used as sort columns in the rewritten ACID query
    */
   default List<FieldSchema> acidSortColumns(org.apache.hadoop.hive.ql.metadata.Table table, Operation operation) {
+    return acidSortColumns(table, operation, false);
+  }
+
+  default List<FieldSchema> acidSortColumns(org.apache.hadoop.hive.ql.metadata.Table table, Operation operation, 
+      boolean emptyOrdering) {
```
[the remark this answers is not in the document]
Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/metadata/HiveStorageHandler.java]: refactored (removed new API) by adding fallback for merge operation in `HiveStorageHandler.acidSortColumns`

**Review by SourabhBadhya — APPROVED** [review]
LGTM +1

**Review by kasakrisz — APPROVED** [review]
